schemas/authoring-guide.md

```diff
@@ -0,0 +1,829 @@
+}
+```
+
+```json
+{
+  "msg": "All good",
+  "tag": "progress",
+  "percent": 0.6,
+  "last": "61626378797a"
+  }
+}
+```
+
+```json
+{
+  "msg": "Ping",
+  "tag": "ping",
+  "ts": 1572935564043,
+  "nonce": "424f524b"
+}
+```
+
+```ipldsch
+type Message struct {
+  msg String
+  payload Payload
+}
+
+type Payload union {
+	| Error "error"
+	| Progress "progress"
+  | Ping "ping"
+} representation inline {
+	discriminantKey "tag"
+}
+
+type Error struct {
+  message String
+}
+
+type Progress struct {
+  percent Float
+  last String
+}
+
+type Ping struct {
+  ts Int
+  nonce String
+}
+```
+
+The interface presented by this Schema is adjusted in comparison to the previous Unions as `Error` is now a Struct with a `message` field.
+
+### Byteprefix Unions for Bytes
+
+A special case union exists for handling Bytes kinds. Where a node contains a byte array (Bytes kind), we may want to discriminate between two different uses of that byte array at the application layer. For example, consider two different encoding schemes where we store a "key" field that is distinct for the each encoding scheme. For practical purposes they are both byte arrays, but at the application layer it helps to have them separated into distinct forms, perhaps so we can make simple assertions about getting the expected key type for the given encoding scheme. There are additional documentation clarity benefits for extracting distinct forms and naming them in a Schema that may factor in to such a decision.
+
+```ipldsch
+type Authorization struct {
+  key PublicKey
+  keyType String
+}
+
+type PublicKey union {
+	| RsaPubkey 0
+	| Ed25519Pubkey 1
+} representation byteprefix
+
+type RsaPubkey bytes
+type Ed25519Pubkey bytes
+```
+
+By declaring a `byteprefix` union, we specify that the first byte of the byte array found at the `key` node of `Authorization` will discriminate which `type` the public key is. That first byte will be sliced off and expected to be either `0x0` or `0x1`, then the remainder of the byte array will be extracted and encapsulated inside either `RsaPubkey` or `Ed25519Pubkey` depending on the discriminator byte.
+
+## Copy
+
+The Copy Schema kind is a special case that provides a mechanism for copying the definition of one named type into a new name. It uses the `=` token after the new type's name followed by name of the type being copied. It is not possible to copy an unnamed (anonymous) type.
+
+```
+type Ping struct {
+  ts Int
+  nonce String
+}
+
+type Pong = Ping
+```
+
+This example is strictly equivalent to the following in terms of the interaction above the Schema layer:
+
+```
+type Ping struct {
+  ts Int
+  nonce String
+}
+
+type Pong struct {
+  ts Int
+  nonce String
+}
+```
+
+The Schema tooling and the reified form of the Schema retains a `copy` kind marker, but tooling that consumes Schemas is expected to treat this marker as an indirection to the named type being copied and copy the entirety of that type's definition to the new name.
```

Comment:
eh, I don't know where to do here, we need a name for this that's concise and specific and "processed form" is neither. We're using "reified" in a few other places, including introduction.md, but it's a word both @warpfork and I have had trouble with I think, perhaps it's this problem of not being a common word.
It arguably is the right word though, https://en.wikipedia.org/wiki/Reification_(computer_science)

If we come up with a better word then I'd be happy to use it, I'm kind of sick of being vague about what this thing is that we stick `.json` after when we spit it out.

Reply:
Oh, I didn't know it's a computer science-y word. I'm fine with it (which can well be, my domain expertise is database indexing and some geo stuff). I tried a normal dictionary as I didn't identify it as a technical term. If I would've used Wikipedia, it would've been easy.

Reply:
Yeah, it's a term I'm somewhat fond of in comp sci literature.

I'd also agree it would be great to find another term for the thing we're refering to here.  "Reified foo" or "reify the frob" is more like a contextual modifier or general verb... it's not great for a title of a significantly important point in the lifecycle of some data.  It works here, but we have this question for this particular piece of data come up in a bunch of places, so a more proper noun for it would be really nice to come up with.

Reply:
AST being a word we often reach for, but it's not a syntax tree. "abstract data tree" was suggested but it doesn't seem to pin anything down - also I don't think "abstract" is even the right word, the thing we're talking about is _more_ Schema than the DSL is.
A closer term is IR, "intermediate representation", that's much more accurate than AST if we're leaning on compile metaphors. "internal representation" is another one I've reached for but I don't mind the simplicity of "intermediate representation". What think you?

Reply:
I see the resonance with "IR" too, yeah.

Some phrasing that emphasizes that this format Is The Thing -- and the DSL by contrast is just window-dressing -- would be ideal...